python/plugins/processing/algs/qgis/Climb.py

```diff
@@ -0,0 +1,246 @@
+            )
+        )
+
+        self.addOutput(
+            QgsProcessingOutputNumber(
+                self.MAXELEVATION,
+                self.tr('Maximum elevation')
+            )
+        )
+
+    def processAlgorithm(self, parameters, context, feedback):
+
+        source = self.parameterAsSource(
+            parameters,
+            self.INPUT,
+            context
+        )
+
+        fcount = source.featureCount()
+
+        hasZ = QgsWkbTypes.hasZ(source.wkbType())
+
+        if not hasZ:
+            feedback.reportError(self.tr('The layer does not have Z values. If you have a DEM, use the Drape algorithm to extract Z values.'))
+            return
+
+        thefields = QgsFields()
+        climbindex = -1
+        descentindex = -1
+        fieldnumber = 0
+
+        # Skip fields with names that are equal to the generated ones
+        for field in source.fields():
+            if str(field.name()) == str(self.CLIMBATTRIBUTE):
+                feedback.pushInfo("Warning: existing " +
+                                  str(self.CLIMBATTRIBUTE) +
+                                  " attribute found and removed")
+                climbindex = fieldnumber
+            elif str(field.name()) == str(self.DESCENTATTRIBUTE):
+                feedback.pushInfo("Warning: existing " +
+                                  str(self.DESCENTATTRIBUTE) +
+                                  " attribute found and removed")
+                descentindex = fieldnumber
+            else:
+                thefields.append(field)
+            fieldnumber = fieldnumber + 1
+
+        # Create new fields for climb and descent
+        thefields.append(QgsField(self.CLIMBATTRIBUTE, QVariant.Double))
+        thefields.append(QgsField(self.DESCENTATTRIBUTE, QVariant.Double))
+        thefields.append(QgsField(self.MINELEVATTRIBUTE, QVariant.Double))
+        thefields.append(QgsField(self.MAXELEVATTRIBUTE, QVariant.Double))
+
+        layerwithz = source
+
+        (sink, dest_id) = self.parameterAsSink(parameters,
+                                               self.OUTPUT,
+                                               context, thefields,
+                                               layerwithz.wkbType(),
+                                               source.sourceCrs())
+
+        # get features from source (with z values)
+        features = layerwithz.getFeatures()
+        totalclimb = 0
+        totaldescent = 0
+        minelevation = float('Infinity')
+        maxelevation = float('-Infinity')
+
+        for current, feature in enumerate(features):
+            # Stop the algorithm if cancelled
+            if feedback.isCanceled():
+                break
+            climb = 0
+            descent = 0
+            minelev = float('Infinity')
+            maxelev = float('-Infinity')
+            # In case of multigeometries we need to do the parts
+            for part in feature.geometry().constParts():
```

Comment:
Better take a copy of the geometry first, `constParts()` directly works on the geometry parts so we need to make sure we keep local reference to a copy of the geometry in order to avoid suddenly having the base geometry modified or deleted (not sure how realistic that is, but taking a copy is on the safe side).